### ISSUE TITLE
useWindowSize debounce options

### ISSUE THREAD

**Issue by lancegliser** [juliencrn/usehooks-ts#213]
I noticed that useWindowSize is hard coded without a method debounce the `setState` call. It causes a significant overhead that I lack a means to offset. I rigged up a super basic optional debounce built on `lodash` to open discussion about how we could handle it. Obviously we'd want to substitute this package's debounce if we use it. There's some conversation to be had about exporting the `setTimeout` in [useDebounce.ts](https://github.com/juliencrn/usehooks-ts/blob/master/src/useDebounce/useDebounce.ts) to ensure it's standard.

hooks/useWindowSize.ts
```ts
import { useState } from "react";
import { useEventListener, useIsomorphicLayoutEffect } from "usehooks-ts";

import { debounce } from "lodash";

interface WindowSize {
  width: number;
  height: number;
}

type OmitFirst<T extends any[]> = T extends [infer A, ...infer R] ? R : never;
type DebounceArgs = OmitFirst<Parameters<typeof debounce>>;

type UseWindowSize = (options?: { debounceArgs?: DebounceArgs }) => WindowSize;
const useWindowSize: UseWindowSize = (options = {}) => {
  const { debounceArgs } = options;
  const [windowSize, setWindowSize] = useState<WindowSize>({
    width: 0,
    height: 0,
  });

  const handleSize = () => {
    console.info(window.innerWidth); // Temporary so you can see it firing more easily.
    setWindowSize({
      width: window.innerWidth,
      height: window.innerHeight,
    });
  };
  const debouncedHandleSize = debounceArgs
    ? debounce(handleSize, ...debounceArgs)
    : handleSize;

  useEventListener("resize", debouncedHandleSize);

  // Set size at the first client-side load
  useIsomorphicLayoutEffect(() => {
    handleSize();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return windowSize;
};

export default useWindowSize;
```

Usage would be dead simple:

```ts
// No debounce
  useWindowSize();
// With debounce
  useWindowSize({
    debounceArgs: [DEBOUNCE_RESIZE_WINDOW_TIMING],
  });
```

**Comment by juliencrn** [comment]
Yes, we should implement the debouncing, I will do it soon. Thanks for sharing that code!

**Comment by bot19** [comment]
Thanks for that. I saw this and was impressed as I was planning on building my own repo of useful hooks I've made throughout the years.

However looking through, I did notice this and it stuck out to me like a huge sore thumb. It's basically incomplete for me to ever implement a window resize hook and not debounce it.

Would love to see this implemented as it's getting close to a year since suggestion. Cheers.
